### ISSUE TITLE
Exception when simulating with many concurrent requests

### ISSUE THREAD

**Issue by laiyhman** [googleads/googleads-java-lib#115]
Hi,

I was trying to simulate many concurrent requests (customerService and managedCustomerService) to observe and understand how the RateLimiter handle the error and i encountered several different types of errors as shown below. Can you help?
A little info about our system and how i do the testing, we're using Apache Camel that consumes the request(1500 of them) in RabbitMQ with 2000 concurrent consumer.

**ArrayIndexOutOfBoundsException**
com.google.api.ads.adwords.extension.ratelimiter.ApiRetryHelper - Failed to Method.getCustomers at exception check: encountered non-retriable java.lang.ArrayIndexOutOfBoundsException, skip retry!

**AxisFault** (probably is caused by the following AxisFault?)
om.google.api.ads.adwords.extension.ratelimiter.ApiRetryHelper - Failed to Method.getCustomers at exception check: encountered non-retriable org.apache.axis.AxisFault, skip retry!

**More than one SoapHeader**
AxisFault
 faultCode: {http://schemas.xmlsoap.org/soap/envelope/}Server
 faultSubcode: 
 faultString: More than one SoapHeader while invoking public abstract java.util.List com.google.ads.api.services.customer.CustomerService.getCustomers() throws com.google.ads.api.services.common.error.ApiException with params [].

**NullPointerException**
java.lang.NullPointerException
	at java.util.concurrent.ConcurrentHashMap.hash(ConcurrentHashMap.java:333)
	at java.util.concurrent.ConcurrentHashMap.get(ConcurrentHashMap.java:988)
	at com.google.common.util.concurrent.AtomicLongMap.get(AtomicLongMap.java:82)
	at com.google.api.ads.adwords.extension.ratelimiter.ApiServicesRetryStrategy.updateAccountWaitTime(ApiServicesRetryStrategy.java:138)
	at com.google.api.ads.adwords.extension.ratelimiter.ApiServicesRetryStrategy.checkRateExceededErrorAndUpdateWaitTime(ApiServicesRetryStrategy.java:196)
	at com.google.api.ads.adwords.extension.ratelimiter.ApiServicesRetryStrategy.shouldRetryOnError(ApiServicesRetryStrategy.java:97)
	at com.google.api.ads.adwords.extension.ratelimiter.ApiRetryHelper.callWithRetries(ApiRetryHelper.java:98)
	at com.google.api.ads.adwords.extension.ratelimiter.ApiRateLimiter.run(ApiRateLimiter.java:60)
	at com.google.api.ads.adwords.extension.ratelimiter.ApiInvocationHandlerWithRateLimiter.handleInvocation(ApiInvocationHandlerWithRateLimiter.java:45)
	at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:84)
	at com.sun.proxy.$Proxy134.getCustomers(Unknown Source)
	at com.mbww.datapoint.google.adwords.Engine.getAndQueueAccounts(Engine.java:153)
	at sun.reflect.GeneratedMethodAccessor237.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.camel.component.bean.MethodInfo.invoke(MethodInfo.java:408)
	at org.apache.camel.component.bean.MethodInfo$1.doProceed(MethodInfo.java:279)
	at org.apache.camel.component.bean.MethodInfo$1.proceed(MethodInfo.java:252)
	at org.apache.camel.component.bean.BeanProcessor.process(BeanProcessor.java:177)
	at org.apache.camel.management.InstrumentationProcessor.process(InstrumentationProcessor.java:77)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:190)
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:460)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:190)
	at org.apache.camel.processor.MulticastProcessor.doProcessSequential(MulticastProcessor.java:668)
	at org.apache.camel.processor.MulticastProcessor.doProcessSequential(MulticastProcessor.java:596)
	at org.apache.camel.processor.MulticastProcessor.process(MulticastProcessor.java:237)
	at org.apache.camel.management.InstrumentationProcessor.process(InstrumentationProcessor.java:77)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:190)
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:121)
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:83)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:190)
	at org.apache.camel.util.AsyncProcessorHelper.process(AsyncProcessorHelper.java:109)
	at org.apache.camel.processor.DelegateAsyncProcessor.process(DelegateAsyncProcessor.java:87)
	at org.apache.camel.component.rabbitmq.RabbitMQConsumer$RabbitConsumer.handleDelivery(RabbitMQConsumer.java:204)
	at com.rabbitmq.client.impl.ConsumerDispatcher$4.run(ConsumerDispatcher.java:121)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:76)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

**INFO**
The following INFO in the log shows that Request ID is null, could this link to the nullpointerException above?
com.google.api.ads.adwords.lib.client.AdWordsServiceClient.requestInfoLogger - Request made: Service: "CustomerServiceInterfacePort" Method: "getCustomers" URL: "https://adwords.google.com/api/adwords/mcm/v201609/CustomerService" Request ID: "null"

**Comment by chenzhuo914** [comment]
Hi laiyhman, thanks for trying the rate-limiter extension and reporting this issue!

NullPointerException: an RateExceededError with "ACCOUNT" scope happens, so RateLimiter is checking the wait time (from previous rate limit handling) of that account. However the input account ID is NULL, leading to this NullPointerException. RateLimiter gets account ID info from the AdWordsSession's clientCustomerId field, meaning that you were calling the AdWords API with an NULL clientCustomerId in AdWordsSession, which is legal in some cases but not expected by RateLimiter for handling RateExceededError - I think it's a bug on RateLimiter and I will look at how to solve it.

Other exceptions don't have detailed call stacks, however they all seem to relate to getCustomers() method. Quoting from the [document](https://developers.google.com/adwords/api/docs/guides/accounts-overview#customerservice):
> If no clientCustomerId is specified in a request, the response will contain multiple entries if more than one account is directly accessible by the authenticated account. If you only want results for a single account, you must specify the clientCustomerId in your request.

So I guess you are calling CustomerService.getCustomers() **without** providing clientCustomerId in AdWordsSession, right? If so, could you check your code logic and see whether you can call CustomerService.getCustomers()  **with** "clientCustomerId" in AdWordsSession (therefore only get one single account info in the result)?

**Comment by laiyhman** [comment]
Hi,

Thanks for the quick reply, the ArrayIndexOutOfBound and AxisFault error i think it could be some mistake on our end which i hope i fixed it, i will run more test and get back if it still happens.
As for the null pointer exception, you're right, we're not providing the clientCustomerId on the first call. Basically this is what we're doing,
`customerService = adWordsServices.get(session, CustomerServiceInterface.class);
// we dont provide the clientCustomerId at the following getCustomer() because we want to get all the accounts that the credential has direct access to,
// and then from there get all the accounts with the acquired clientCustomerId. We have a scenario that actual user will not provide any id 
// and just get whatever acccount that they can access to
Customer[] customers = customerService.getCustomers();  
com.google.api.ads.adwords.axis.v201609.cm.Selector selector = new com.google.api.ads.adwords.axis.v201609.cm.Selector();
String [] fields = new String[]{"CustomerId", "CanManageClients","Name", "CurrencyCode", "DateTimeZone","AccountLabels"};
selector.setFields(fields);

if (customers!=null) // accounts that the credential has direct access to
for(int i=0; i < customers.length ;i++)
{
// get all the accounts from the "parent" account
	//loop through all accounts attached with this login credential and from that main account get all managed customer accounts
	session.setClientCustomerId(customers[i].getCustomerId().toString());
	managedCustomerService = adWordsServices.get(session, ManagedCustomerServiceInterface.class);
	ManagedCustomerPage managedCustomerPage = managedCustomerService.get(selector);
	manageAccounts.addAll(Lists.newArrayList(managedCustomerPage.getEntries()));
}
`

**Comment by chenzhuo914** [comment]
I think your use case makes total sense for your business logic.

So, I will make a fix in RateLimiter. Meanwhile, you can workaround this issue by making the first call (CustomerService.getCustomers() **without** providing clientCustomerId) **WITHOUT** RateLimtier extension, then making subsequent calls (all with clientCustomerId) with RateLimiter extension.

**Comment by laiyhman** [comment]
That sounds great!

Thanks!

**Comment by laiyhman** [comment]
Hi,

I have another question, is there a way to simulate daily download limit exceeded error? We cannot use our current account to do so as it's a production account.
And how does the RateLimiter work in this case, does it wait 24 hours until it tries to download again? Is there a way to extract any info so that we can get back to our UI to inform about the waiting so that user will not wait several hours and not seeing anything?

Thanks.

**Comment by laiyhman** [comment]
Hi,

Another question,

I see the following logs in the aw-ratelimiter.log only if I use the RateLimiter without setting clientCustomerId like mentioned earlier. If i remove the RateLimiter extension from the code that does not set clientCustomerId then i won't see any log in the aw-ratelimiter.log. Does it mean that if the RateLimiter does it job without any issue then there wont be anything on the aw-ratelimiter.log?
Or how do i know that there's rateexceeded issue and that the extension has handled it properly?

***********
2017-05-16 20:54:15,626  INFO ApiServicesRetryStrategy: 
***********
Encountered RateExceededError: scope=ACCOUNT, seconds=30.

**Comment by chenzhuo914** [comment]
Regarding "daily download limit", if your developer token is in basic access level, it can only download [1000 reports per day](https://developers.google.com/adwords/api/docs/guides/rate-limits#operational_limit_based_on_access_level). Otherwise the rate limit is dynamic and you don't know when it will happen. An alternative way is to monitor the RateExceededError, and if the retryAfterSeconds is 86400 (which 24 hours) then it probably reaches daily limit. In this case, RateLimit would stop retry and fail immediately (with default config value ApiServicesRetryStrategy. MAX_WAIT_TIME_ON_RATE_EXCEEDED_ERROR_DEFAULT=86400).

Regarding logging, when RateLimiter is running it will append log messages to aw-ratelimiter.log, and you can control log granularity by modifying [log4j.properties](https://github.com/googleads/googleads-java-lib/blob/master/extensions/adwords_rate_limiter/src/main/resources/log4j.properties). Whenever RateExceededError happens, RateLimiter would print the info log as you saw ("Encountered RateExceededError: scope=ACCOUNT, seconds=30.") **before** handling it, [here](https://github.com/googleads/googleads-java-lib/blob/master/extensions/adwords_rate_limiter/src/main/java/com/google/api/ads/adwords/extension/ratelimiter/ApiServicesRetryStrategy.java#L185) is the relevant code.

**Comment by laiyhman** [comment]
Hi,

Do you have a sample code on the error that will be thrown when the daily download limit is reached so that i can handle it on our end?

**Comment by chenzhuo914** [comment]
I think you don't need to focus too much on "daily limit", which is only applicable to basic level developer token. Suppose you make some hacky code logic around it, and later your developer token gets upgraded to standard access, are you sure such code logic won't cause some side effect?

Instead, you should just handle rate limit properly according to this [guide doc](https://developers.google.com/adwords/api/docs/guides/rate-limits). We don't have sample code, which would be language specific, but a simplified logic could be:

```
if some AdWords API call throws ApiException
  iterate all the errors inside that ApiException
    if an error is of type RateExceededError
      get scope and retryAfterSeconds from that RateExceededError
      if scope is "ACCOUNT"
        pause all AdWords API calls towards that account for at least retryAfterSeconds
      else if scope is "DEVELOPER"
        pause all AdWords API calls using that developer token for at least retryAfterSeconds
      end if
    end if
  end iteration
end if
```

**Comment by laiyhman** [comment]
Hi,

Based on your previous comments 

_if the retryAfterSeconds is 86400 (which 24 hours) then it probably reaches daily limit. In this case, RateLimit would stop retry and fail immediately_

If i understand it correctly, the RateLimiter would not know whether the rateexceedederror is because of the queries per second thing or it's related to the daily download limit error, except that we can assume that it's a daily limit issue if the retry seconds is 24 hours. And RateLimiter would simply fail the operation immediately. So when that happens our application needs to be aware of the situation that the daily limit has exceeded so that we can inform the user the task fail because of daily limit is done.
Since the RateLimiter handle the situation, thus i'm trying to understand the behavior and if the RateLimiter would simply throw the same rateexceedederror where my application can catch and handle it something like your sample pseudo code above and return to user?
We can't test that situation because we're only on Basic level now. However I'm setting up a test account and hopefully i can test the daily limit with it.
But still appreciate if you can answer my questions, thank you for your patience :)

**Comment by chenzhuo914** [comment]
No, RateLimiter doesn't differentiate QPS or daily limit, and it's better not to make such assumptions because you don't have to.

If you wonder about RateLimiter behavior when the wait time is too long, please explore [its code](https://github.com/googleads/googleads-java-lib/tree/master/extensions/adwords_rate_limiter/src/main/java/com/google/api/ads/adwords/extension/ratelimiter) (it's open-source isn't it?)

Hint: pay more attention on these functions
- ApiRetryHelper.callWithRetries()
- ApiServicesRetryStrategy.calcWaitTimeBeforeCall() / ApiReportingRetryStrategy.calcWaitTimeBeforeCall()

Also, you mentioned downloading reports just now. RateLimiter provides 2 ApiRetryStrategy implementations:
- ApiServicesRetryStrategy for handling service calls (such as CustomerService, ManagedCustomerService)
- ApiReportingRetryStrategy for handing report downloading calls

You can see their respective settings (with default values which you can override) in the [Getting Started section](https://github.com/googleads/googleads-java-lib/tree/master/extensions/adwords_rate_limiter#getting-started) of README.

**Comment by laiyhman** [comment]
Thanks, i'll take a look again

**Comment by laiyhman** [comment]
Hi,

`System.setProperty(ApiServicesRetryStrategy.MAX_WAIT_TIME_ON_RATE_EXCEEDED_ERROR_PROPERTY, "86400")`

The field ApiServicesRetryStrategy.MAX_WAIT_TIME_ON_RATE_EXCEEDED_ERROR_PROPERTY is not visible for external package

I'll just use the value directly for now

**Comment by chenzhuo914** [comment]
It's a known issue and will be fixed in next release.

Btw, those settings have default values. You only call System.setProperty() if you want to override those default values.

**Comment by nwbirnie** [comment]
Fixed in the latest version (0.1.1 released Monday 15th January 2018).

Specifically ApiReportingRetryStrategy.java now declares the property variables as public so you can reference them directly.

Thank you very much for reporting this issue. We really appreciate your feedback and hope that you enjoy using the component.
